streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamsTopologyGraph.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.kstream.internals;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The builder of the Streams Intermediate DAG used to make
+ * optimizations of a Kafka Streams topology and sub-topologies
+ */
+public abstract class StreamsTopologyGraph {
+
+    public static final String TOPOLOGY_ROOT = "root";
+    protected final StreamsGraphNode root = new StreamsGraphNode(TOPOLOGY_ROOT, TopologyNodeType.TOPOLOGY_PARENT);
+
+    protected final Map<StreamsGraphNode, Set<StreamsGraphNode>> repartitioningNodeToRepartitioned = new HashMap<>();
+    protected final Map<StreamsGraphNode, StreamsGraphNode> stateStoreNodeToSinkNodes = new HashMap<>();
+    protected final Map<String, StreamsGraphNode> nameToGraphNode = new HashMap<>();
+
+    public StreamsTopologyGraph() {
+        nameToGraphNode.put(TOPOLOGY_ROOT, root);
+    }
+
+    /**
+     * Add a StreamsGraphNode to the graph
+     *
+     * @param metadata        the metadata node to add
+     */
+    abstract void addNode(StreamsGraphNode metadata);
+
+    /**
+     * Retrieves the root node of the DAG of metadata for the topology
+     * allowing access and traversal of entire topology metadata
+     *
+     * @return StreamsGraphNode root
+     */
+    public StreamsGraphNode root() {
+        return root;
+    }
+
+    /**
+     * Used for hints when a node in the topology triggers a repartition and the repartition flag
+     * is propagated down through the descendant nodes of the topology.  This can be used to help make an
+     * optimization where the triggering node does an eager "through" operation and the child nodes can ignore
+     * the need to repartition.
+     *
+     * @return Map&lt;StreamsGraphNode, Set&lt;StreamGraphNode&gt;&gt;
+     */
+    public Map<StreamsGraphNode, Set<StreamsGraphNode>> getRepartitioningNodeToRepartitioned() {
```

Comment:
following 3 methods don't seem to be used

Reply:
during this first phase they aren't but when we do optimizations they will be.